### PR TITLE
[ticket/13192] Add method for generating valid user page links based on mod_rewrite

### PR DESCRIPTION
**Pull request by marc1706**
Ticket: https://tracker.phpbb.com/browse/PHPBB3-13192

### CODE REVIEW

File: phpBB/phpbb/path_helper.php [hunk 1]
```diff
@@ -445,4 +445,38 @@ public function append_url_params($url, $new_params, $is_amp = true)
...
+		{
+			// Remove 'app.php/' from the page, when rewrite is enabled.
+			// Treat app.php as a reserved file name and remove on mod rewrite
+			// even if it might not be in the phpBB root.
+			if ($mod_rewrite && ($app_position = strpos($page, 'app.' . $this->php_ext . '/')) !== false)
+			{
+				$page = substr($page, 0, $app_position) . substr($page, $app_position + strlen('app.' . $this->php_ext . '/'));
+			}
+
+			// Remove preceding slashes from page name and append to root path
```
Comment (marc1706) [inline phpBB/phpbb/path_helper.php]: That's the same thing basically? Do you prefer your version?

Comment (nickvergessen) [inline phpBB/phpbb/path_helper.php]: well you do not modify root path ;) so you do not append it to the root path ;)

Comment (marc1706) [inline phpBB/phpbb/path_helper.php]: Of course I'm not modifying the root path. I'm appending the modified page name to the root path.

Comment (michaelcullum) [inline phpBB/phpbb/path_helper.php]: This is correct although 'append to **the** root path'

@nickvergessen You are `prepending the root path to the page name` or `appending the page name to the root path`.

Comment (marc1706) [inline phpBB/phpbb/path_helper.php]: The "the" was left out on purpose.

File: tests/path_helper/path_helper_test.php [hunk 2]
```diff
@@ -421,4 +421,27 @@ public function test_get_web_root_path_from_ajax_referer($referer_url, $board_ur
 	{
 		$this->assertEquals($this->phpbb_root_path . $expected, $this->path_helper->get_web_root_path_from_ajax_referer($referer_url, $board_url));
 	}
+
+	public function data_get_valid_page()
+	{
+		return array(
+			// array( current page , mod_rewrite setting , expected output )
+			array('index', true, 'index'),
+			array('index', false, 'index'),
```
Comment (Nicofuma) [inline tests/path_helper/path_helper_test.php]: it's still mixed

Comment (marc1706) [inline tests/path_helper/path_helper_test.php]: Hm, seems like I didn't push the changes O.o

**Comment by marc1706** [comment]
@nickvergessen Updated test cases

File: phpBB/phpbb/path_helper.php [hunk 3]
```diff
@@ -445,4 +445,38 @@ public function append_url_params($url, $new_params, $is_amp = true)
...
+			// Remove 'app.php/' from the page, when rewrite is enabled.
+			// Treat app.php as a reserved file name and remove on mod rewrite
+			// even if it might not be in the phpBB root.
+			if ($mod_rewrite && ($app_position = strpos($page, 'app.' . $this->php_ext . '/')) !== false)
+			{
+				$page = substr($page, 0, $app_position) . substr($page, $app_position + strlen('app.' . $this->php_ext . '/'));
+			}
+
+			// Remove preceding slashes from page name and prepend root path
+			$page = $this->get_phpbb_root_path() . preg_replace('@^(?:([\\/\\\])?)@', '', $page);
```
Comment (nickvergessen) [inline phpBB/phpbb/path_helper.php]: `$page = $this->get_phpbb_root_path() . ltrim($page, '/');`

**Comment by marc1706** [comment]
@nickvergessen updated

File: phpBB/includes/functions.php [hunk 4]
```diff
@@ -2719,8 +2700,8 @@ function confirm_box($check, $title = '', $hidden = '', $html_body = 'confirm_bo
 	}
 
 	// re-add sid / transform & to &amp; for user->page (user->page is always using &)
-	$use_page = ($u_action) ? $phpbb_root_path . $u_action : $phpbb_root_path . str_replace('&', '&amp;', $user->page['page']);
-	$u_action = reapply_sid($use_page);
+	$use_page = ($u_action) ? $u_action : str_replace('&', '&amp;', $user->page['page']);
+	$u_action = reapply_sid($phpbb_path_helper->get_valid_page($use_page, $config['enable_mod_rewrite']));
```
Comment (nickvergessen) [inline phpBB/includes/functions.php]: We must only to that when `$u_action` is not used.
`$u_action` is a parameter of `confirm_box()` that allows using custom URLs,
and we should not modify this further?

Comment (marc1706) [inline phpBB/includes/functions.php]: We should only apply the session ID if u_action is not used? That doesn't make any sense. What would reapplying the session ID do if we're not using it? And why shouldn't we make sure to apply the sid to the u_action path that might have been modified using the user page?

**Comment by marc1706** [comment]
@bantu @naderman Could you maybe review this one to make sure confirm_box() and build_url() aren't broken in any way by these changes?

**Pull request merged**